src/sdam/topology.ts

```diff
@@ -459,40 +460,44 @@ export class Topology extends TypedEventEmitter<TopologyEvents> {
       }
     }
 
-    const exitWithError = (error: Error) =>
-      callback ? callback(error) : this.emit(Topology.ERROR, error);
+    const exitWithError = (error: Error) => {
+      this.emit(Topology.ERROR, error);
+      throw error;
+    };
 
     const readPreference = options.readPreference ?? ReadPreference.primary;
     const selectServerOptions = { operationName: 'ping', ...options };
-    this.selectServer(
-      readPreferenceServerSelector(readPreference),
-      selectServerOptions,
-      (err, server) => {
-        if (err) {
-          this.close();
-          return exitWithError(err);
-        }
-
-        const skipPingOnConnect = this.s.options[Symbol.for('@@mdb.skipPingOnConnect')] === true;
-        if (!skipPingOnConnect && server && this.s.credentials) {
-          server.command(ns('admin.$cmd'), { ping: 1 }, {}).then(() => {
-            stateTransition(this, STATE_CONNECTED);
-            this.emit(Topology.OPEN, this);
-            this.emit(Topology.CONNECT, this);
-
-            callback?.(undefined, this);
-          }, exitWithError);
+    try {
+      const server = await this.selectServer(
+        readPreferenceServerSelector(readPreference),
+        selectServerOptions
+      );
 
-          return;
+      const skipPingOnConnect = this.s.options[Symbol.for('@@mdb.skipPingOnConnect')] === true;
+      if (!skipPingOnConnect && server && this.s.credentials) {
+        try {
+          await server.command(ns('admin.$cmd'), { ping: 1 }, {});
+          stateTransition(this, STATE_CONNECTED);
+          this.emit(Topology.OPEN, this);
+          this.emit(Topology.CONNECT, this);
+
+          return this;
+        } catch (error) {
+          exitWithError(error);
```

Comment:
Previously, we conditionally switched between calling a callback or emitting error dependent on whether a callback was provided. Ever since we wrapped this in promisify we were always providing a callback. Is it correct that we now always emit an error? If not then I think we can get rid of the try catch and let things throw on their own.